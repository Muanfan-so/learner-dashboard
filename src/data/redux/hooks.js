import { useSelector } from 'react-redux';

import { actions as appActions } from './app/reducer';
import appSelectors from './app/selectors';
import requestSelectors from './requests/selectors';

const { courseCard } = appSelectors;

export const usePageNumber = () => useSelector(appSelectors.pageNumber);
export const useEmailConfirmationData = () => useSelector(appSelectors.emailConfirmation);
export const useEnterpriseDashboardData = () => useSelector(appSelectors.enterpriseDashboard);
export const usePlatformSettingsData = () => useSelector(appSelectors.platformSettings);
// suggested courses is max at 3 at the moment.
export const useSuggestedCoursesData = () => useSelector(appSelectors.suggestedCourses).slice(0, 3);
export const useRecommendedCoursesData = () => useSelector(appSelectors.recommendedCourses);
export const useSelectSessionModalData = () => useSelector(appSelectors.selectSessionModal);

export const useHasCourses = () => useSelector(appSelectors.hasCourses);
export const useHasAvailableDashboards = () => useSelector(appSelectors.hasAvailableDashboards);
export const useCurrentCourseList = (opts) => useSelector(
  state => appSelectors.currentList(state, opts),
);
export const useShowSelectSessionModal = () => useSelector(appSelectors.showSelectSessionModal);

// eslint-disable-next-line
export const useCourseCardData = (selector) => (cardId) => useSelector(
  (state) => selector(state, cardId),
);

export const useCardCertificateData = useCourseCardData(courseCard.certificate);
export const useCardCourseData = useCourseCardData(courseCard.course);
export const useCardCourseRunData = useCourseCardData(courseCard.courseRun);
export const useCardEnrollmentData = useCourseCardData(courseCard.enrollment);
export const useCardEntitlementData = useCourseCardData(courseCard.entitlement);
export const useCardGradeData = useCourseCardData(courseCard.gradeData);
export const useCardProviderData = useCourseCardData(courseCard.courseProvider);
export const useCardRelatedProgramsData = useCourseCardData(courseCard.relatedPrograms);

export const useUpdateSelectSessionModalCallback = (dispatch, cardId) => () => dispatch(
  appActions.updateSelectSessionModal(cardId),
);

export const useMasqueradeData = () => useSelector(requestSelectors.masquerade);

export const useIsPendingRequest = (requestName) => useSelector(requestSelectors.isPending(requestName));
