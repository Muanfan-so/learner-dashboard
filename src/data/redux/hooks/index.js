import * as app from './app';
import * as requests from './requests';

export * from './app';
export { requests };

export default { app, requests };
