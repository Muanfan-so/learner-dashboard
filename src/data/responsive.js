import { useWindowSize, breakpoints } from '@edx/paragon';
import { StrictDict } from 'utils';

const createResponsive = (fn) => () => {
  const { width } = useWindowSize();
  return fn(width);
};

export const isDesktopSize = createResponsive(
  (width) => width >= breakpoints.large.minWidth,
);

// Beware of using isPadSize() for styling. When the screen resizes pass the
// breakpoint, the screen will flash between the two styles. This is why it is
// recommened isDesktopSize() or isMobileSize() be used instead.
export const isIpadSize = createResponsive(
  (width) => width >= breakpoints.medium.minWidth && width < breakpoints.large.minWidth,
);

export const isMobileSize = createResponsive(
  (width) => width < breakpoints.medium.minWidth,
);

export default StrictDict({
  isDesktopSize,
  isIpadSize,
  isMobileSize,
});
