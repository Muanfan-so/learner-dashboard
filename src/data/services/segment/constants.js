import { StrictDict } from 'utils';

export const events = StrictDict({
  courseEnroll: 'courseEnroll',
  entitlementUnenroll: 'entitlementUnenroll',
  sessionChange: 'sessionChange',
  unenrollReason: 'unenrollReason',
  upgradeCourse: 'upgradeCourse',
  searchCourse: 'searchCourse',
});

export const eventNames = StrictDict({
  [events.courseEnroll]: 'edx.bi.user.program-details.enrollment',
  [events.upgradeCourse]: 'learner_home.course_card.upgrade',
  [events.entitlementUnenroll]: 'entitlement_unenrollment_reason.selected',
  [events.sessionChange]: ({ action }) => `course-dashboard.${action}-session`, // 'switch', 'new', 'leave'
  [events.unenrollReason]: 'unenrollment_reason.selected',
  [events.searchCourse]: 'learner_home.widget.search_course',
});

export const trackingCategory = 'learner-home';

export const pageViewEvent = { category: trackingCategory };
