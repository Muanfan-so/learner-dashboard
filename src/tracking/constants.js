import { StrictDict } from 'utils';

export const categories = StrictDict({
  dashboard: 'dashboard',
  upgrade: 'upgrade',
  userEngagement: 'user-engagement',
  searchButton: 'search_button',
});

export const events = StrictDict({
  enterCourseClicked: 'enterCourseClicked',
  courseImageClicked: 'courseImageClicked',
  courseTitleClicked: 'courseTitleClicked',
  courseOptionsDropdownClicked: 'courseOptionsDropdownClicked',
  upgradeButtonClicked: 'upgradeButtonClicked',
  upgradeButtonClickedEnrollment: 'upgradeButtonClickedEnrollment',
  upgradeButtonClickedUpsell: 'upgradeButtonClickedUpsell',
  shareClicked: 'shareClicked',
  userSettingsChanged: 'userSettingsChanged',
  newSession: 'newSession',
  switchSession: 'switchSession',
  leaveSession: 'leaveSession',
  unenrollReason: 'unenrollReason',
  entitlementUnenrollReason: 'entitlementUnenrollReason',
  enterpriseDashboardModalOpened: 'enterpriseDashboardModalOpened',
  enterpriseDashboardModalCTAClicked: 'enterpriseDashboardModalCTAClicked',
  enterpriseDashboardModalClosed: 'enterpriseDashboardModalClosed',
});

const learnerPortal = 'edx.ui.enterprise.lms.dashboard.learner_portal_modal';

export const eventNames = StrictDict({
  enterCourseClicked: 'edx.bi.dashboard.enter_course.clicked',
  courseImageClicked: 'edx.bi.dashboard.course_image.clicked',
  courseTitleClicked: 'edx.bi.dashboard.course_title.clicked',
  courseOptionsDropdownClicked: 'edx.bi.dashboard.course_options_dropdown.clicked',
  upgradeButtonClicked: 'edx.bi.dashboard.upgrade_button.clicked',
  upgradeButtonClickedEnrollment: 'edx.course.enrollment.upgrade.clicked',
  upgradeButtonClickedUpsell: 'edx.bi.ecommerce.upsell_links_clicked',
  shareClicked: 'edx.course.share_clicked',
  userSettingsChanged: 'edx.user.settings.changed',
  newSession: 'course-dashboard.new-session',
  switchSession: 'course-dashboard.switch-session',
  leaveSession: 'course-dashboard.leave-session',
  unenrollReason: 'unenrollment_reason.selected',
  entitlementUnenrollReason: 'entitlement_unenrollment_reason.selected',
  enterpriseDashboardModalOpened: `${learnerPortal}.opened`,
  enterpriseDashboardModalCTAClicked: `${learnerPortal}.dashboard_cta.clicked`,
  enterpriseDashboardModalClosed: `${learnerPortal}.closed`,
  findCoursesClicked: 'edx.bi.dashboard.find_courses_button.clicked',
});

export const appName = 'learner-home';

export default eventNames;
