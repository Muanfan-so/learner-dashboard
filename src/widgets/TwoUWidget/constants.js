/* eslint-disable quotes */
import { StrictDict } from 'utils';

const countryCodeUS = 'US';

const merchandisingKeys = StrictDict({
  execed: 'execed',
  execedDescription: 'execedDescription',
  masters: 'masters',
  mastersDescription: 'mastersDescription',
  bachelors: 'bachelors',
  bachelorsDescription: 'bachelorsDescription',
  bootcamps: 'bootcamps',
  bootcampsDescription: 'bootcampsDescription',
});

const messages = StrictDict({
  [merchandisingKeys.execed]: {
    id: 'TwoUWidget.execed',
    defaultMessage: 'Executive Education',
    description: 'The title for the link to the executive education page.',
  },
  [merchandisingKeys.execedDescription]: {
    id: 'TwoUWidget.execedDescription',
    defaultMessage: 'Short courses to develop leadership skills',
    description: 'The description for the link to the executive education page.',
  },
  [merchandisingKeys.masters]: {
    id: 'TwoUWidget.masters',
    defaultMessage: `Master's Degrees`,
    description: 'The title for the link to the masters page.',
  },
  [merchandisingKeys.mastersDescription]: {
    id: 'TwoUWidget.mastersDescription',
    defaultMessage: 'Online degree programs from top universities',
    description: 'The description for the link to the masters page.',
  },
  [merchandisingKeys.bachelors]: {
    id: 'TwoUWidget.bachelors',
    defaultMessage: `Bachelor's Degrees`,
    description: 'The title for the link to the bachelors page.',
  },
  [merchandisingKeys.bachelorsDescription]: {
    id: 'TwoUWidget.bachelorsDescription',
    defaultMessage: 'Begin or complete a degree; fully online',
    description: 'The description for the link to the bachelors page.',
  },
  [merchandisingKeys.bootcamps]: {
    id: 'TwoUWidget.bootcamps',
    defaultMessage: 'Boot Camps',
    description: 'The title for the link to the boot camps page.',
  },
  [merchandisingKeys.bootcampsDescription]: {
    id: 'TwoUWidget.bootcampsDescription',
    defaultMessage: 'Intensive, hands-on, project-based training',
    description: 'The description for the link to the boot camps page.',
  },
});

const merchandisingItems = [
  {
    key: merchandisingKeys.execed,
    title: messages[merchandisingKeys.execed],
    description: messages[merchandisingKeys.execedDescription],
    itemURL: 'https://www.edx.org/executive-education?vanguards_click=execed',
  },
  {
    key: merchandisingKeys.masters,
    title: messages[merchandisingKeys.masters],
    description: messages[merchandisingKeys.mastersDescription],
    itemURL: 'https://www.edx.org/masters?vanguards_click=masters',
  },
  {
    key: merchandisingKeys.bachelors,
    title: messages[merchandisingKeys.bachelors],
    description: messages[merchandisingKeys.bachelorsDescription],
    itemURL: 'https://www.edx.org/bachelors?vanguards_click=bachelors',
  },
  {
    key: merchandisingKeys.bootcamps,
    title: messages[merchandisingKeys.bootcamps],
    description: messages[merchandisingKeys.bootcampsDescription],
    itemURL: 'https://www.edx.org/boot-camps?vanguards_click=bootcamps',
  },
];

const getMerchandisingItems = (countryCode) => (
  countryCode === countryCodeUS ? merchandisingItems : merchandisingItems.slice(0, 3)
);

export { getMerchandisingItems, countryCodeUS };
