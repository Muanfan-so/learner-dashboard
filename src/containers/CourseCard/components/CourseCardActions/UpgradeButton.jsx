import React from 'react';
import PropTypes from 'prop-types';

import { Locked } from '@edx/paragon/icons';
import { useIntl } from '@edx/frontend-platform/i18n';

import { hooks } from 'data/redux';
import ActionButton from './ActionButton';
import messages from './messages';

export const UpgradeButton = ({ cardId }) => {
  const { upgradeUrl } = hooks.useCardCourseRunData(cardId);
  const { canUpgrade } = hooks.useCardEnrollmentData(cardId);
  const { isMasquerading } = hooks.useMasqueradeData();
  const { formatMessage } = useIntl();
  const isEnabled = (!isMasquerading && canUpgrade);
  return (
    <ActionButton
      iconBefore={Locked}
      variant="outline-primary"
      disabled={!isEnabled}
      {...isEnabled && { as: 'a', href: upgradeUrl }}
    >
      {formatMessage(messages.upgrade)}
    </ActionButton>
  );
};
UpgradeButton.propTypes = {
  cardId: PropTypes.string.isRequired,
};
export default UpgradeButton;
