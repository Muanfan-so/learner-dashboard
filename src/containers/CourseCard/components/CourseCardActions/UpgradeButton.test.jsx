import { shallow } from 'enzyme';

import { htmlProps } from 'data/constants/htmlKeys';
import { hooks } from 'data/redux';
import UpgradeButton from './UpgradeButton';

jest.mock('data/redux', () => ({
  hooks: {
    useMasqueradeData: jest.fn(() => ({ isMasquerading: false })),
    useCardCourseRunData: jest.fn(),
    useCardEnrollmentData: jest.fn(() => ({ canUpgrade: true })),
  },
}));
jest.mock('./ActionButton', () => 'ActionButton');

describe('UpgradeButton', () => {
  const props = {
    cardId: 'cardId',
  };
  const upgradeUrl = 'upgradeUrl';
  hooks.useCardCourseRunData.mockReturnValue({ upgradeUrl });
  describe('snapshot', () => {
    test('can upgrade', () => {
      const wrapper = shallow(<UpgradeButton {...props} />);
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.prop(htmlProps.disabled)).toEqual(false);
    });
    test('cannot upgrade', () => {
      hooks.useCardEnrollmentData.mockReturnValueOnce({ canUpgrade: false });
      const wrapper = shallow(<UpgradeButton {...props} />);
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.prop(htmlProps.disabled)).toEqual(true);
    });
    test('masquerading', () => {
      hooks.useMasqueradeData.mockReturnValueOnce({ isMasquerading: true });
      const wrapper = shallow(<UpgradeButton {...props} />);
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.prop(htmlProps.disabled)).toEqual(true);
    });
  });
});
