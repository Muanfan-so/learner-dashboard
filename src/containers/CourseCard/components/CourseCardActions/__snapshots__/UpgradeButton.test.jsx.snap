// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpgradeButton snapshot can upgrade 1`] = `
<ActionButton
  as="a"
  disabled={false}
  href="upgradeUrl"
  iconBefore={[MockFunction icons.Locked]}
  onClick={[MockFunction trackUpgradeClick]}
  variant="outline-primary"
>
  Upgrade
</ActionButton>
`;

exports[`UpgradeButton snapshot cannot upgrade 1`] = `
<ActionButton
  disabled={true}
  iconBefore={[MockFunction icons.Locked]}
  onClick={[MockFunction trackUpgradeClick]}
  variant="outline-primary"
>
  Upgrade
</ActionButton>
`;

exports[`UpgradeButton snapshot masquerading 1`] = `
<ActionButton
  disabled={true}
  iconBefore={[MockFunction icons.Locked]}
  onClick={[MockFunction trackUpgradeClick]}
  variant="outline-primary"
>
  Upgrade
</ActionButton>
`;
